File: sdk/js/src/schemas/bool.ts
import type { ParseContext, SchemaNode } from "../types.js";
import { BaseSchema } from "./base.js";
import { ISSUE_CODES } from "../issue-codes.js";
import { describeType } from "../util.js";

export class BoolSchema extends BaseSchema<boolean, boolean> {
  _getCoercionTarget(): string {
    return "bool";
  }

  _validate(input: unknown, ctx: ParseContext): unknown {
    if (typeof input !== "boolean") {
      ctx.issues.push({
        code: ISSUE_CODES.INVALID_TYPE,
        message: `Expected boolean, received ${describeType(input)}`,
        path: [...ctx.path],
        expected: "bool",
        received: describeType(input),
      });
      return undefined;
    }
    return input;
  }

  _toNode(): SchemaNode {
    const node: SchemaNode = { kind: "bool" } as SchemaNode;
    this._addDefault(node);
    return node;
  }
}

File: sdk/python/src/anyvali/schemas/bool.py
"""Bool schema."""

from __future__ import annotations

from typing import Any

from ..issue_codes import INVALID_TYPE
from .base import BaseSchema, ValidationContext, _anyvali_type_name


class BoolSchema(BaseSchema[bool]):
    """Schema for boolean values."""

    def __init__(self) -> None:
        super().__init__()

    def _validate(self, input: Any, ctx: ValidationContext) -> Any:
        if not isinstance(input, bool):
            received = _anyvali_type_name(input)
            ctx.add_issue(INVALID_TYPE, f"Expected boolean, received {received}", expected="boolean", received=received)
            return None
        return input

    def _to_node(self) -> dict[str, Any]:
        return self._add_common_node_fields({"kind": "bool"})
